libafl_frida/src/alloc.rs

```diff
@@ -0,0 +1,408 @@
+use hashbrown::HashMap;
+use libafl::bolts::os::walk_self_maps;
+use nix::{
+    libc::memset,
+    sys::mman::{mmap, MapFlags, ProtFlags},
+};
+
+use backtrace::Backtrace;
+#[cfg(unix)]
+use libc::{sysconf, _SC_PAGESIZE};
+use rangemap::RangeSet;
+use serde::{Deserialize, Serialize};
+use std::{ffi::c_void, io};
+
+use crate::{
+    asan_errors::{AsanError, AsanErrors},
+    FridaOptions,
+};
+
+pub(crate) struct Allocator {
+    options: FridaOptions,
+    page_size: usize,
+    shadow_offset: usize,
+    shadow_bit: usize,
+    pre_allocated_shadow: bool,
+    allocations: HashMap<usize, AllocationMetadata>,
+    shadow_pages: RangeSet<usize>,
+    allocation_queue: HashMap<usize, Vec<AllocationMetadata>>,
+    largest_allocation: usize,
+    base_mapping_addr: usize,
+    current_mapping_addr: usize,
+}
+
+macro_rules! map_to_shadow {
+    ($self:expr, $address:expr) => {
+        (($address >> 3) + $self.shadow_offset) & ((1 << ($self.shadow_bit + 1)) - 1)
+    };
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub(crate) struct AllocationMetadata {
+    pub address: usize,
+    pub size: usize,
+    pub actual_size: usize,
+    pub allocation_site_backtrace: Option<Backtrace>,
+    pub release_site_backtrace: Option<Backtrace>,
+    pub freed: bool,
+    pub is_malloc_zero: bool,
+}
+
+impl Allocator {
+    pub fn new(options: FridaOptions) -> Self {
+        let ret = unsafe { sysconf(_SC_PAGESIZE) };
+        if ret < 0 {
+            panic!("Failed to read pagesize {:?}", io::Error::last_os_error());
+        }
+        #[allow(clippy::cast_sign_loss)]
+        let page_size = ret as usize;
+        // probe to find a usable shadow bit:
+        let mut shadow_bit: usize = 0;
+        for try_shadow_bit in &[46usize, 36usize] {
+            let addr: usize = 1 << try_shadow_bit;
+            if unsafe {
+                mmap(
+                    addr as *mut c_void,
+                    page_size,
+                    ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+                    MapFlags::MAP_PRIVATE
+                        | MapFlags::MAP_ANONYMOUS
+                        | MapFlags::MAP_FIXED
+                        | MapFlags::MAP_NORESERVE,
+                    -1,
+                    0,
+                )
+            }
+            .is_ok()
+            {
+                shadow_bit = *try_shadow_bit;
+                break;
+            }
+        }
+        assert!(shadow_bit != 0);
```

Comment:
Andrea added some debug assert macro crate that gets optimized out on release

Reply:
I need run time verification that I picked a working `shadow_bit`.